src/historywork/BatchWork.h

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+#pragma once
+
+#include "work/Work.h"
+#include <medida/meter.h>
+#include <medida/metrics_registry.h>
+
+namespace medida
+{
+class Meter;
+}
+
+namespace stellar
+{
+
+class TmpDir;
+
+class BatchWork : public Work
+{
+    /* This class performs batch download (and optionally verification) of
+     *files. This class supplies a mechanism for parallel file verification and
+     *throttles the workers; Child classes must supply iteration methods, as
+     *well as actual download + verify work. This class only acts as a
+     *commander.
+     **/
+  public:
+    BatchWork(Application& app, WorkParent& parent, std::string const& type,
+              TmpDir const& downloadDir);
+    virtual ~BatchWork() override;
+    virtual void onReset() override;
+    virtual void notify(std::string const& child) override;
+
+    virtual bool hasNext() = 0;
+    virtual std::string getNextDownloadFilename() = 0;
+    virtual std::string assignChildWork(std::string ft) = 0;
```

Comment:
`assignChildWork` does not make sense to me - I would expect a function that returns or modifies some `Work` instead of hijacking the assignment of a child from a filename